### PR TITLE
Reimplement IVF_PQ index.

### PR DESCRIPTION
**Pull request by eddyxu**
Using PQ quantizer globally to build a global codebook.

### CODE REVIEW

**Comment by eddyxu** [comment]
Without refine step

```
./metrics.py ../../../rust/sift_1m.lance -k 10
nprobs: 1, refine=None recall@10=0.500
nprobs: 6, refine=None recall@10=0.680
nprobs: 11, refine=None recall@10=0.600
nprobs: 16, refine=None recall@10=0.690
nprobs: 21, refine=None recall@10=0.910
nprobs: 26, refine=None recall@10=0.700
nprobs: 31, refine=None recall@10=0.740
nprobs: 36, refine=None recall@10=0.700
nprobs: 41, refine=None recall@10=0.780
nprobs: 46, refine=None recall@10=0.800
```

With `refine_factor=2`

```
./metrics.py ../../../rust/sift_1m.lance -k 10 --refine 2
nprobs: 1, refine=2 recall@10=0.850
nprobs: 6, refine=2 recall@10=0.720
nprobs: 11, refine=2 recall@10=0.850
nprobs: 16, refine=2 recall@10=0.750
nprobs: 21, refine=2 recall@10=0.820
nprobs: 26, refine=2 recall@10=0.740
nprobs: 31, refine=2 recall@10=0.740
nprobs: 36, refine=2 recall@10=0.840
nprobs: 41, refine=2 recall@10=0.810
nprobs: 46, refine=2 recall@10=0.840
```

Refine factor == 5

```
❯ ./metrics.py ../../../rust/sift_1m.lance -k 10 --refine 5
nprobs: 1, refine=5 recall@10=0.630
nprobs: 6, refine=5 recall@10=0.930
nprobs: 11, refine=5 recall@10=0.910
nprobs: 16, refine=5 recall@10=0.990
nprobs: 21, refine=5 recall@10=0.960
nprobs: 26, refine=5 recall@10=0.960
nprobs: 31, refine=5 recall@10=0.990
nprobs: 36, refine=5 recall@10=0.970
nprobs: 41, refine=5 recall@10=0.960
nprobs: 46, refine=5 recall@10=0.950
```

File: rust/src/index/vector/pq.rs [hunk 1]
```diff
@@ -281,6 +246,30 @@ impl ProductQuantizer {
     }
 }
 
+impl From<&pb::Pq> for ProductQuantizer {
+    fn from(proto: &pb::Pq) -> Self {
+        Self {
+            nbits: proto.num_bits,
```
Comment (changhiskhan) [inline rust/src/index/vector/pq.rs:252]: nit: don't feel strongly at all but should we just call this `num_bits`

File: rust/src/index/vector/pq.rs [hunk 2]
```diff
@@ -26,14 +26,15 @@ use arrow_ord::sort::sort_to_indices;
...
+use crate::index::pb;
 use crate::index::vector::kmeans::train_kmeans;
 use crate::io::object_reader::{read_fixed_stride_array, ObjectReader};
 use crate::Result;
 use crate::{arrow::*, utils::distance::l2_distance};
 
 /// Product Quantization Index.
 ///
-pub struct PQIndex {
+pub struct PQIndex<'a> {
```
Comment (changhiskhan) [inline rust/src/index/vector/pq.rs:37]: i still don't understand lifetimes. why is this needed here?

File: rust/src/index/vector/ivf.rs [hunk 3]
```diff
@@ -339,46 +353,50 @@ impl Ivf {
...
+                let residual = tokio::task::spawn_blocking(move || {
+                    compute_residual(
+                        centorids.clone(),
+                        as_fixed_size_list_array(vector.as_ref()),
+                        as_primitive_array(partition_ids.as_ref()),
+                    )
+                })
+                .await??;
+                let residual_schema = Arc::new(ArrowSchema::new(vec![
+                    ArrowField::new("__residual_vector", residual.data_type().clone(), false),
```
Comment (changhiskhan) [inline rust/src/index/vector/ivf.rs:382]: make this a constant as well?

**Review by changhiskhan — COMMENTED** [review]
just some minor questions

File: rust/src/encodings/plain.rs [hunk 4]
```diff
@@ -78,13 +78,15 @@ impl<'a> PlainEncoder<'a> {
         }
     }
 
+    /// Encode primitive values.
     async fn encode_primitive(&mut self, array: &dyn Array) -> Result<usize> {
         let offset = self.writer.tell();
-        let data = array.data().buffers()[0].as_ref();
+        let data = array.data().buffers()[0].as_slice();
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply rust/src/encodings/plain.rs:84]: `as_slice` sets the offset / length correctly if the buffer was actually a "slice" of another buffer

https://docs.rs/arrow-buffer/32.0.0/src/arrow_buffer/buffer/immutable.rs.html#139

File: rust/src/index/vector/pq.rs (hunk 2 above)
Comment (eddyxu) [reply rust/src/index/vector/pq.rs:37]: PQIndex now does not own the `ProductQuantizer`, but just has a reference to the `ProductQuantizer`.  So PQIndex needs to make sure `ProductQuantizer` is alive.

File: rust/src/index/vector/pq.rs (hunk 1 above)
Comment (eddyxu) [reply rust/src/index/vector/pq.rs:252]: k, will do.

File: rust/src/io/reader.rs [hunk 5]
```diff
@@ -261,14 +261,25 @@ async fn read_batch(
         .await?;
     let mut batch = RecordBatch::try_new(Arc::new(schema.into()), arrs)?;
     if with_row_id {
+        let ids_in_batch: Vec<i32> = match params {
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply rust/src/io/reader.rs:264]: This calculates the index from the correct form of the input parameters.

For example, if the batch starts with 100, and it tries to take `[5, 22, 13]`the row id would be `105, 122, 123`

File: protos/index.proto [hunk 6]
```diff
@@ -64,6 +64,12 @@ message PQ {
 
   // Number of sub vectors.
   uint32 num_sub_vectors = 2;
+
+  // Vector dimension
+  uint32 dimension = 3;
+
+  // Codebook. `dimension * 2 ^ num_bits` of float32s.
+  repeated float codebook = 4;
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply protos/index.proto:72]: For `1028D,  nbits = 8`, this can be 1MB. 

I guess I can make this into an array, but it requires 2 I/O (one for reading this protobuf, and the second one to read the actual codebook), which has data dependency so that we can not parallelize them.  But it gives a chance to read the array in parallel if the array is too big.

File: rust/src/index/vector/ivf.rs (hunk 3 above)
Comment (eddyxu) [reply rust/src/index/vector/ivf.rs:382]: done

**Review by changhiskhan — APPROVED** [review]
